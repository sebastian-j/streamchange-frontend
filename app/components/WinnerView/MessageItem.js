import React from 'react';
import PropTypes from 'prop-types';

function MessageItem(props) {
  const dt = new Date(props.date);
  const convertedDate = `${dt.getHours()}:${
    dt.getMinutes() < 10 ? '0' : ''
  }${dt.getMinutes()}:${dt.getSeconds() < 10 ? '0' : ''}${dt.getSeconds()}`;
  return (
    <li className="winner-message">
      <span>{convertedDate}</span>
      <span>{props.text}</span>
    </li>
  );
}

MessageItem.propTypes = {
  date: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
};

export default MessageItem;
