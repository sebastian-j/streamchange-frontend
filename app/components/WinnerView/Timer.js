import React from 'react';
import styled from 'styled-components';

const Span = styled.span`
  color: gray;
  margin-left: auto;
`;

export default class Timer extends React.Component {
  constructor(props) {
    super(props);
    this.state = { display: '', interval: null };
  }

  componentDidMount() {
    this.setState({ display: '0:00' });
    const start = new Date().getTime();
    const t = setInterval(() => {
      const now = new Date().getTime();
      const distance = now - start;
      const minutes = Math.floor((distance % (1000 * 60 * 60)) / (1000 * 60));
      const sec = Math.floor((distance % (1000 * 60)) / 1000);
      this.setState({ display: `${minutes}:${sec < 10 ? '0' : ''}${sec}` });
    }, 1000);
    this.setState({ interval: t });
  }

  componentWillUnmount() {
    clearInterval(this.state.interval);
  }

  render() {
    return <Span>{this.state.display}</Span>;
  }
}
