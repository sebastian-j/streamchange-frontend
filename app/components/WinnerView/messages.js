/*
 * WinnerView Messages
 *
 * This contains all the text for the WinnerView component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.components.WinnerView';

export default defineMessages({
  panelTitle: {
    id: `${scope}.title`,
    defaultMessage: 'Winner',
  },
  exitBtn: {
    id: `${scope}.exitBtn`,
    defaultMessage: 'Return',
  },
  loading: {
    id: `${scope}.loading`,
    defaultMessage: 'Loading...',
  },
  openChannel: {
    id: `${scope}.openChannel`,
    defaultMessage: 'Open channel page',
  },
  prize: {
    id: `${scope}.prize`,
    defaultMessage: 'Prize',
  },
  saveBtn: {
    id: `${scope}.saveBtn`,
    defaultMessage: 'Save and return',
  },
  notSubscribed: {
    id: `${scope}.SubStatus.notSubscribed`,
    defaultMessage: 'Not subscribed',
  },
  subscriberPrivate: {
    id: `${scope}.SubStatus.privateSub`,
    defaultMessage: 'Subscription list private',
  },
  subscriberFrom: {
    id: `${scope}.SubStatus.subscriberFrom`,
    defaultMessage: 'Subscriber from',
  },
  subscriberUnknown: {
    id: `${scope}.SubStatus.unknownSubscriber`,
    defaultMessage: 'Subscription status unknown',
  },
});
