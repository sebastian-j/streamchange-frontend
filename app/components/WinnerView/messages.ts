/*
 * WinnerView Messages
 *
 * This contains all the text for the WinnerView component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.components.WinnerView';

export default defineMessages({
  pageTitle: {
    id: `${scope}.pageTitle`,
    defaultMessage: 'wins the draw',
  },
  panelTitle: {
    id: `${scope}.title`,
    defaultMessage: 'Winner',
  },
  exitBtn: {
    id: `${scope}.exitBtn`,
    defaultMessage: 'Return',
  },
  loading: {
    id: `${scope}.loading`,
    defaultMessage: 'Loading...',
  },
  openChannel: {
    id: `${scope}.openChannel`,
    defaultMessage: 'Open channel page',
  },
  prize: {
    id: `${scope}.prize`,
    defaultMessage: 'Prize',
  },
  saveBtn: {
    id: `${scope}.saveBtn`,
    defaultMessage: 'Save and return',
  },
  replayBtn: {
    id: `${scope}.replayBtn`,
    defaultMessage: 'Repeat the draw',
  },
  replayBtnTooltip: {
    id: `${scope}.replayBtnTooltip`,
    defaultMessage:
      'Repeats the draw immediately using the same settings. Current winner won\'t be saved in history.',
  },
  notSubscribed: {
    id: `${scope}.SubStatus.notSubscribed`,
    defaultMessage: 'Not subscribed',
  },
  subscriberPrivate: {
    id: `${scope}.SubStatus.privateSub`,
    defaultMessage: 'Subscription list private',
  },
  subscriberFrom: {
    id: `${scope}.SubStatus.subscriberFrom`,
    defaultMessage: 'Subscriber from',
  },
  subscriberUnknown: {
    id: `${scope}.SubStatus.unknownSubscriber`,
    defaultMessage: 'Subscription status unknown',
  },
});
