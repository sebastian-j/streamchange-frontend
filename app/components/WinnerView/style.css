.winner-outer {
    display: flex;
    flex-direction: row;
    padding: 10px;
}

.winner-logo {
    height: 70px;
    width: 70px;
    margin-right: 10px;
}

.winner-title {
    font-size: 20px;
}

.winner-sub-status {
    font-size: 0.9rem;
    padding-bottom: 3px;
}

.winner-sub-status.sub-true {
    font-weight: bold;
    color: #007703;
}
.winner-btn {
    background: none;
    border: 1px solid #0059a3;
    color: #0059a3;
    border-radius: 4px;
    padding: 3px 5px;
    text-decoration: none;
}

.winner-btn:hover {
    background-color: #0059a3;
    color: white;
}

.winner-message-list {
    overflow-y: auto;
    list-style: none;
    padding: 0;
}

.winner-message {
    margin-bottom: 5px;
}
.winner-message span:nth-child(1) {
    color: grey;
    margin-right: 10px;
}