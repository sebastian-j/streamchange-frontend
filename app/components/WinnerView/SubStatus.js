import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';

import messages from './messages';
import { makeSelectStreamInfo } from '../../containers/HomePage/selectors';
import RelativeDate from '../RelativeDate';

const Subscribed = styled.span`
  color: ${(props) => props.theme.subStatusPositive};
  font-weight: bold;
  font-size: 0.9rem;
  padding-bottom: 3px;
`;

const NotSubscribed = styled.span`
  color: ${(props) => props.theme.subStatusNegative};
  font-weight: bold;
  font-size: 0.9rem;
  padding-bottom: 3px;
`;

const PrivateSubs = styled.span`
  color: ${(props) => props.theme.staticTextColor};
  font-size: 0.9rem;
  max-width: 60%;
  padding-bottom: 3px;
`;

const SubStatus = (props) => {
  const [status, setStatus] = useState(null);
  const [subscriberFrom, setSubscriberFrom] = useState(null);

  const checkStatus = () => {
    axios
      .get(
        `https://www.googleapis.com/youtube/v3/subscriptions?part=snippet&channelId=${props.id}&forChannelId=${props.streamInfo.ownerId}&fields=items/snippet/publishedAt&key=${props.apiKey}`,
      )
      .then((res) => {
        if (res.data.items.length > 0) {
          setStatus('true');
          setSubscriberFrom(res.data.items[0].snippet.publishedAt);
        } else {
          setStatus('false');
        }
      })
      .catch((err) => {
        if (
          err.response &&
          err.response.data &&
          err.response.data.error.errors[0].reason === 'subscriptionForbidden'
        ) {
          setStatus('private');
        } else {
          setStatus('error');
        }
      });
  };

  useEffect(() => {
    checkStatus();
  }, []);
  if (!status) {
    return <PrivateSubs>...</PrivateSubs>;
  }
  if (status === 'true') {
    return (
      <Subscribed>
        <FormattedMessage {...messages.subscriberFrom} />
        <RelativeDate ISO8601Date={subscriberFrom} />
      </Subscribed>
    );
  }
  if (status === 'false') {
    return (
      <NotSubscribed>
        <FormattedMessage {...messages.notSubscribed} />
      </NotSubscribed>
    );
  }
  if (status === 'private') {
    return (
      <PrivateSubs>
        <FormattedMessage {...messages.subscriberPrivate} />
      </PrivateSubs>
    );
  }
  return (
    <PrivateSubs>
      <FormattedMessage {...messages.subscriberUnknown} />
    </PrivateSubs>
  );
};

SubStatus.propTypes = {
  apiKey: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  streamInfo: PropTypes.object.isRequired,
};

const mapStateToProps = createStructuredSelector({
  streamInfo: makeSelectStreamInfo(),
});

export default connect(mapStateToProps, null)(SubStatus);
