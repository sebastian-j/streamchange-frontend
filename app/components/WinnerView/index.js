import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import db from '../YoutubeWorker/db';
import MessageItem from './MessageItem';
import RelativeDate from '../RelativeDate';
import './style.css';

export default class WinnerView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      imageUrl: null,
      title: null,
      messages: [],
      interval: null,
      subscriberFrom: null,
      subscriberStatus: null,
    };
    this.getMessages = this.getMessages.bind(this);
    this.checkSubStatus = this.checkSubStatus.bind(this);
  }

  getMessages() {
    const authorId = this.props.id;
    db.table('messages')
      .filter(function(message) {
        return message.authorId === authorId;
      })
      .toArray()
      .then(items => {
        this.setState({ messages: items });
      });
  }

  checkSubStatus() {
    axios
      .get(
        `https://www.googleapis.com/youtube/v3/subscriptions?part=snippet&channelId=${
          this.props.id
        }&forChannelId=${this.props.ownerId}&key=${this.props.apiKey}`,
      )
      .then(res => {
        if (res.data.items.length > 0) {
          this.setState({
            subscriberFrom: res.data.items[0].snippet.publishedAt,
            subscriberStatus: 'true',
          });
        } else {
          this.setState({ subscriberStatus: 'false' });
        }
      })
      .catch(err => {
        if (
          err.response &&
          err.response.data &&
          err.response.data.error.errors[0].reason === 'subscriptionForbidden'
        ) {
          this.setState({ subscriberStatus: 'private' });
        } else {
          this.setState({ subscriberStatus: 'error' });
        }
      });
  }

  componentDidMount() {
    const userId = this.props.id;
    db.table('users')
      .filter(function(user) {
        return user.id === userId;
      })
      .toArray()
      .then(items => {
        this.setState({ imageUrl: items[0].imageUrl, title: items[0].title });
      });
    this.checkSubStatus();
    this.getMessages();
    this.setState({ interval: setInterval(this.getMessages.bind(this), 3000) });
  }

  componentWillUnmount() {
    clearInterval(this.state.interval);
  }

  render() {
    if (!this.state.title || !this.state.subscriberStatus) {
      return (
        <div className="gv-column flex-column">
          <h2 className="column-title">Zwycięzca</h2>
          <span>Ładowanie...</span>
          <button
            className="winner-btn"
            onClick={this.props.onClose}
            type="button"
          >
            Powrót
          </button>
        </div>
      );
    }
    return (
      <div className="gv-column flex-column">
        <h2 className="column-title">Zwycięzca</h2>
        <div className="winner-outer">
          <img className="winner-logo" alt="logo" src={this.state.imageUrl} />
          <div className="flex-column">
            <span className="winner-title">{this.state.title}</span>
            {this.state.subscriberStatus === 'true' && (
              <span className="winner-sub-status sub-true">
                Subskrybuje od&nbsp;
                <RelativeDate ISO8601Date={this.state.subscriberFrom} />
              </span>
            )}
            {this.state.subscriberStatus === 'false' && (
              <span className="winner-sub-status">Nie subskrybuje</span>
            )}
            {this.state.subscriberStatus === 'private' && (
              <span className="winner-sub-status">Subskrypcje prywatne</span>
            )}
            <a
              href={`https://www.youtube.com/channel/${this.props.id}`}
              target="_blank"
              className="winner-btn"
            >
              Przejdź na kanał
            </a>
          </div>
        </div>
        <ul className="winner-message-list">
          {this.state.messages.map(item => (
            <MessageItem date={item.publishedAt} text={item.displayText} />
          ))}
        </ul>
        <button
          className="winner-btn"
          onClick={this.props.onClose}
          type="button"
        >
          Powrót
        </button>
      </div>
    );
  }
}

WinnerView.propTypes = {
  apiKey: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  ownerId: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
};
