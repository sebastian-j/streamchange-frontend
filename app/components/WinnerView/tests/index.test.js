import React from 'react';
import { Provider } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import configureStore from 'redux-mock-store';

import WinnerView from '../index';

const renderer = new ShallowRenderer();
const mockStore = configureStore([]);

describe('<WinnerView />', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      isOpen: false,
    });
  });
  it('should render and match the snapshot', () => {
    renderer.render(
      <Provider store={store}>
        <WinnerView apiKey="key" id="id" ownerId="owner" onClose={() => 0} />
      </Provider>,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
