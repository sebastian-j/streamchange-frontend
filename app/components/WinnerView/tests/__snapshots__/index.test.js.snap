// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WinnerView /> should render and match the snapshot 1`] = `
<ForwardRef>
  <ForwardRef>
    <FormattedMessage
      defaultMessage="Winner"
      id="streamchange.components.WinnerView.title"
      values={Object {}}
    />
  </ForwardRef>
  <span>
    <FormattedMessage
      defaultMessage="Loading..."
      id="streamchange.components.WinnerView.loading"
      values={Object {}}
    />
  </span>
  <ForwardRef
    onClick={[Function]}
    type="button"
  >
    <FormattedMessage
      defaultMessage="Return"
      id="streamchange.components.WinnerView.exitBtn"
      values={Object {}}
    />
  </ForwardRef>
</ForwardRef>
`;
